package.json

```diff
@@ -83,6 +83,6 @@
     "preversion": "npm run build && (git diff --quiet || git commit -am build)"
   },
   "engines": {
-    "node": ">= 8.16.2"
+    "node": "^10 || ^12 || >=14"
```

Comment:
I wonder if it would be better to remove the `engines` field as no one seems committed to keeping it up to date.

Reply:
I'd prefer to keep it so that downstream users would be at least notified by NPM if they are using an unsupported version. Maybe we can automate it in a script as part of the build process?. I'll have a think about the lightest approach. Potentially parsing https://github.com/nodejs/Release/blob/master/schedule.json, and have a play tomorrow.

Reply:
> notified by NPM if they are using an unsupported version

In the [package.json spec currently](https://docs.npmjs.com/cli/v6/configuring-npm/package-json/#engines) it is advisory at this time. Pay particular notice to `engine-strict` and not `engineStrict`. Knowing if a package could be advised on install that it is incompatible would be nice, if *npm* gets those and related issues in those issues fixed. We use the package.json advisory since we're "top level" but I also mention to other devs to read CHANGELOGs assuming a project has one.

See also:
* https://github.com/npm/cli/issues/692 *(hover for issue label status)*
* https://github.com/npm/cli/issues/786 *(hover for issue label status)*

I do read deprecation warnings too in case *node* changes their mind.

Reply:
We should definitely keep `engines`. Its not just a signal to consumers but also anyone who wants to contribute to the project.

Even though npm doesn't respect `engines`, yarn does.

Reply:
> Its not just a signal to consumers but also anyone who wants to contribute to the project.

The problems with engines as I see them:

- No one looks at it to make sure their PR complies with it.
- No one reviewing PRs makes sure they are complying with that field.
- No one updates it when versions of node pass their end of life.

Therefore it is usually out of date and not very useful to anyone.

If there is a way to automate updating it and testing for active node versions, or if someone is willing to commit to keeping it up to date and review PRs to make sure they comply, I am all for having. If there is no way to automate it or no one step forward to maintain it, it just seems like a burden for us developers and sometimes useful to users.

Reply:
Vercel deployments use it, so its is implicitly tested in the sense that Node.js 14 is selected based on `engines` when building our docs and api.

Maybe we just need a test to compare `engines` to GH CI to make sure we're testing what is defined there.

Reply:
> Vercel deployments use it

But they would use the same version if we didn't have it.

> Maybe we just need a test to compare `engines` to GH CI to make sure we're testing what is defined there.

How do we do that? Currently we only test `lts` and `latest` node versions so something could break node v10 and we wouldn't know it.

Reply:
Ah, when you said you only support lts and latest, I thought you meant *all* current LTS versions of node (10 and 12) as well as latest (14) rather than just the most recent LTS

Reply:
I feel like the `engines` field, at best, is a crutch. If someone wants to know which versions we support we should put that in the docs but relying on the `engines` field to be correct seems ridiculous if it is not necessarily going to be up to date.

"When looking for the time it is better to have not found a clock at all then to have found one that is incorrect." -- Someone Smart

Reply:
Let's correct the clock then 👍 

> But they would use the same version if we didn't have it.

No, they would use the version assigned in the project settings. See [the docs](https://vercel.com/docs/runtimes#official-runtimes/node-js/node-js-version).

Reply:
I think @styfle had something close to the right idea.
We could do the following if you want to support all LTS + latest:

1. Create a script to fetch the current latest node release versions & filter for LTS and current
2. Run these latest versions in a semver.satisfies call against the package.json `engines` field.
3. Run this script as part of CI. If a new version of node comes out then we'll be instantly notified and can update `engines` in that PR who's branch failed.

Reply:
> Ah, when you said you only support lts and latest, I thought you meant all current LTS versions of node

Realistically we support whatever version someone is willing to maintain. If someone wants to submit a PR that fixes v2 for node v8 I would happily accept it, as long as it doesn't make things slower or break things for newer node versions.

Reply:
> Whenever a new Project is created, the latest Node.js LTS version available on Vercel at that time is selected for it.

Great so we don't need the engines field according to the docs.

> Let's correct the clock then 👍

But who is going to make sure the clock stays correct?  and what is the correct version? as anyone tested it on node v10.0.0? if not than `>=10` could be incorrect.

Reply:
Maybe we can make use of Renovate https://docs.renovatebot.com/node/

Reply:
> Maybe we can make use of Renovate

I would be ok with Renovate. I use it on most of my other projects to auto-merge dependency updates.

> Does it hurt to add it to CI?

Ya if we can get it automated than I would be ok with keeping it.

> unless you mean the original 10

Tecnically `v10.0.0` is in the range `>=10` so someone looking in engines or using yarn would be correct in assuming `v10.0.0` was a supported node version.

Reply:
If we aren't testing against Node 10 (looks like we're not) then we probably shouldn't say we support it.

Also since 10 is reaching EOL soon lets just say `>=12`.

My main point is that our CI needs needs to match what we expect. If we drop support for a version of Node.js, that would be a major semver.

Reply:
> Also since 10 is reaching EOL soon lets just say `>=12`.

So we should support v12.0.0 and v12.0.1 ...? otherwise if the `engines` field is set to `>=12` than it is about as useful as a broken clock.

> If we drop support for a version of Node.js, that would be a major semver.

I agree, but as I said above we really support whatever someone is willing to maintain.

Reply:
> My main point is that our CI needs needs to match what we expect.

I think it does. I expect marked to work with `lts` and `latest` versions of node. If we could do something like

```json
  engines: {
    "node": "lts || latest"
  }
```

I would be all for it.

Reply:
> So we should support v12.0.0 and v12.0.1

Yes, of course. Since Node.js follows semver we can assume code that works on 12.0.0 will work on 12.0.1.

> "node": "lts || latest"

I think this is backwards.

You're trying to make `engines` match our current CI but I think we need to make our CI match `engines` by being explicit (test both 12.x and 14.x which are both LTS and probably 15.x which is stable).

https://nodejs.org/en/about/releases/

Reply:
> Yes, of course. Since Node.js follows semver we can assume code that works on 12.0.0 will work on 12.0.1.

Not true, For example: `fs.rm(dir)` would work in node v14.14.0 but not in v14.13.0. The `fs.rm` function was added in v14.14.0

So just because something works for latest version in a major version does not mean it will work for every version in that major version. Semver only guarantees the opposite. If something works in v14.0.0 it should work in every later version with the same major version number. 

And every fix could be a breaking change for someone's [workflow](https://xkcd.com/1172/)

Reply:
> You're trying to make `engines` match our current CI

Yes, I am trying to make `engines` match the versions known to work with marked (because they are the only ones we test). The problem with testing specific versions or supporting specific versions is that they get out of date really fast and everyone forgets about them in smaller projects like marked.

Reply:
I understand how useful `engines` can be **if it is accurate**. I just don't believe anyone is willing to keep it accurate enough to be useful in marked.

Reply:
I'd suggest:
- setting the range in `engines` to start with the first LTS version of the major. So something like `"^10.13 || ^12.13  || ^14.15  || >=15"`. Features are not back ported anymore after a release reaches LTS so you have a consistent basis.
- If you want to be extra safe, add 10.13.0, 12.13.0 and 14.15.0 to the build matrix. That way you'll notice when a new features breaks an older version and can bump the `engines` field whenever the build matrix is changed - this is also easy to review.

As an aside, strong vote from my end to bump the major whenever adjusting the list of supported Node or Browser version - it makes things a lot easier for downstream consumers.

Reply:
@realityking The issue I have with setting the `engines` field is it is very difficult to keep it accurate for a project as small as marked where there are often weeks or even months between changes.

This is a community project, no person or company has any commercial interest in marked. So keeping the `engines` field up to date seems like a waste of time that provides no benefit because it will rarely be accurate and because as I stated in https://github.com/markedjs/marked/pull/1938#discussion_r575383861 we support any version of node that someone is willing to maintain.